src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs

```diff
@@ -19,8 +19,23 @@ internal static class DelegateHelpers
         // with the Reflection.Emit statics below.
         private static class DynamicDelegateLightup
         {
-            public static Func<Type, Func<object?[], object?>, Delegate> CreateObjectArrayDelegate { get; }
-                = CreateObjectArrayDelegateInternal();
+            public static Func<Type, Func<object?[], object?>, Delegate> CreateObjectArrayDelegate
+            {
+                get
+                {
+                    // CreateObjectArrayDelegateInternal is only supported with NativeAOT which always expects CanEmitObjectArrayDelegate to be false.
+                    // Additionally, this check guards static constructor of trying to resolve 'Internal.Runtime.Augments.DynamicDelegateAugments'
+                    // on runtimes which do not support this private API, which would otherwise cause TypeInitializationException.
+                    if (!CanEmitObjectArrayDelegate)
+                    {
+                        return CreateObjectArrayDelegateInternal();
+                    }
+                    else
+                    {
+                        throw new System.NotImplementedException();
+                    }
+                }
+            }
```

Comment:
This doesn't do the same thing anymore.  before you had `public static string MyProp {get; } = MyMethod()` and now you have `public static string MyProp => MyMethod()`.  The first one only calls `MyMethod` once from the class constructor.  The new version calls `MyMethod` every time.

Can't you change `CreateObjectArrayDelegateInternal()` instead like this:

```csharp
  if (!CanEmitObjectArrayDelegate) {
    return Type.GetType(...).GetMethod(...).CreateDelegate();
} else {
  throw new System.NotImplementedException();
}
```

That way the delegate is still only initialized once

Reply:
Actually shouldn't it be more like 

```csharp
if (!CanEmitObjectArrayDelegate) {
  return Type.GetType(...).GetMethod(...).CreateDelegate();
} else {
  return new Func<Type, Func<object?[], object?>, Delegate>((_x, _y) => throw new NotImplementedException(...);
}
```

So it's not the static constructor that throws.  Instead the returned delegate throws when it is called.  And we expect that it will never be called.

Reply:
👍

Reply:
@lambdageek @eerhardt thank you both for suggestions!

I went with Aleksey's approach as when trimming is enabled the end result is identical to what we had prior to this change (since `CanEmitObjectArrayDelegate` will get trimmed away`)

do you agree?

Reply:
The thing I dislike about this approach is that we are already guarding against `if (CanEmitObjectArrayDelegate)` in the call to this code. So it looks like a redundant check that can be removed.

We already have:

https://github.com/dotnet/runtime/blob/8986374ea73f1e5b1977dd0058fbdf591936408a/src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs#L42-L52

Which is the only place that calls into `DynamicDelegateLightup`. So it looks redundant to _yet again_ check for `if (!CanEmitObjectArrayDelegate)` here.

Reply:
Converting this code to use UnsafeAccessor should address this duplication and make the code simpler overall.